### ISSUE TITLE
ItemsAdder called the blockBreak Event twice.

### ISSUE THREAD

**Issue by Rastart21** [PluginBugs/Issues-ItemsAdder#3139]
### Terms

- [X] I'm using the very latest version of ItemsAdder and its dependencies.
- [X] I am sure this is a bug and it is not caused by a misconfiguration or by another plugin.
- [X] I already searched on this [Github page](https://github.com/PluginBugs/Issues-ItemsAdder/issues) to check if the same issue was already reported.
- [X] I already searched on the [plugin wiki](https://itemsadder.devs.beer/) to know if a solution is already known.
- [X] I already searched on the [forums](https://forum.devs.beer/) to check if anyone already has a solution for this.

### Discord tag (optional)

rastart_

### What happened?

So, when i make a quest using BetonQuest with the objective is to break custom block generated by ItemsAdder it counted as 2 when breaking 1 block. I already asked the dev of BetonQuest and his response is to asked you about this "ItemsAdder called the blockBreak Event twice. I think you need to report this to ItemsAdder"

### Steps to reproduce the issue

I'll send you the zip file necesarry to reproduce this issue. Just drag&drop ItemsAdder and BetonQuest to your server directory.

1. Reload IA resourcepack and config.
2. Reload BetonQuest.
3. Place 10 Block named Test (the block texture is the same as my profile).
4. run `/bq event <your_name> test-ia_block.obj1`.
5. Break the block in survival (This bug only happen when you break the block in survival and not in creative).
6. you can see in chat that when breaking 1 block the quest counted as 2.

*please do this on fresh install of itemsadder because when you have another block and added my block it will have different value and needed to be changed in BetonQuest.

### Server version

This server is running Paper version git-Paper-194 (MC: 1.20.1) (Implementing API version 1.20.1-R0.1-SNAPSHOT) (Git: 1b1c230)
You are 2 version(s) behind
Download the new version at: https://papermc.io/downloads/paper
Previous version: git-Paper-511 (MC: 1.19.4)

### ItemsAdder Version

ItemsAdder 3.6.2-beta-r2

### ProtocolLib Version

ProtocolLib 5.1.1-SNAPSHOT-673

### LoneLibs Version

LoneLibs 1.0.27

### FULL server log

https://mclo.gs/DVcznfX

### Error (optional)

```shell
[22:04:54] [Server thread/ERROR]: [BetonQuest] Tracker Break: 
java.lang.Exception: Tracker stack trace
	at org.betonquest.betonquest.objectives.BlockObjective.onBlockBreak(BlockObjective.java:67) ~[BetonQuest (1).jar:?]
	at com.destroystokyo.paper.event.executor.asm.generated.GeneratedEventExecutor836.execute(Unknown Source) ~[?:?]
	at org.bukkit.plugin.EventExecutor$2.execute(EventExecutor.java:77) ~[paper-api-1.20.1-R0.1-SNAPSHOT.jar:?]
	at co.aikar.timings.TimedEventExecutor.execute(TimedEventExecutor.java:81) ~[paper-api-1.20.1-R0.1-SNAPSHOT.jar:git-Paper-194]
	at org.bukkit.plugin.RegisteredListener.callEvent(RegisteredListener.java:70) ~[paper-api-1.20.1-R0.1-SNAPSHOT.jar:?]
	at io.papermc.paper.plugin.manager.PaperEventManager.callEvent(PaperEventManager.java:54) ~[paper-1.20.1.jar:git-Paper-194]
	at io.papermc.paper.plugin.manager.PaperPluginManagerImpl.callEvent(PaperPluginManagerImpl.java:126) ~[paper-1.20.1.jar:git-Paper-194]
	at org.bukkit.plugin.SimplePluginManager.callEvent(SimplePluginManager.java:615) ~[paper-api-1.20.1-R0.1-SNAPSHOT.jar:?]
	at ia.m.jS.a(SourceFile:401) ~[ItemsAdder_3.5.1.jar:?]
	at ia.m.cu.c(SourceFile:190) ~[ItemsAdder_3.5.1.jar:?]
	at org.bukkit.craftbukkit.v1_20_R1.scheduler.CraftTask.run(CraftTask.java:101) ~[paper-1.20.1.jar:git-Paper-194]
	at org.bukkit.craftbukkit.v1_20_R1.scheduler.CraftScheduler.mainThreadHeartbeat(CraftScheduler.java:480) ~[paper-1.20.1.jar:git-Paper-194]
	at net.minecraft.server.MinecraftServer.tickChildren(MinecraftServer.java:1479) ~[paper-1.20.1.jar:git-Paper-194]
	at net.minecraft.server.dedicated.DedicatedServer.tickChildren(DedicatedServer.java:446) ~[paper-1.20.1.jar:git-Paper-194]
	at net.minecraft.server.MinecraftServer.tickServer(MinecraftServer.java:1393) ~[paper-1.20.1.jar:git-Paper-194]
	at net.minecraft.server.MinecraftServer.runServer(MinecraftServer.java:1170) ~[paper-1.20.1.jar:git-Paper-194]
	at net.minecraft.server.MinecraftServer.lambda$spin$0(MinecraftServer.java:317) ~[paper-1.20.1.jar:git-Paper-194]
	at java.lang.Thread.run(Thread.java:833) ~[?:?]
[22:04:54] [Server thread/ERROR]: [BetonQuest] Tracker Break: 
java.lang.Exception: Tracker stack trace
	at org.betonquest.betonquest.objectives.BlockObjective.onBlockBreak(BlockObjective.java:67) ~[BetonQuest (1).jar:?]
	at com.destroystokyo.paper.event.executor.asm.generated.GeneratedEventExecutor836.execute(Unknown Source) ~[?:?]
	at org.bukkit.plugin.EventExecutor$2.execute(EventExecutor.java:77) ~[paper-api-1.20.1-R0.1-SNAPSHOT.jar:?]
	at co.aikar.timings.TimedEventExecutor.execute(TimedEventExecutor.java:81) ~[paper-api-1.20.1-R0.1-SNAPSHOT.jar:git-Paper-194]
	at org.bukkit.plugin.RegisteredListener.callEvent(RegisteredListener.java:70) ~[paper-api-1.20.1-R0.1-SNAPSHOT.jar:?]
	at io.papermc.paper.plugin.manager.PaperEventManager.callEvent(PaperEventManager.java:54) ~[paper-1.20.1.jar:git-Paper-194]
	at io.papermc.paper.plugin.manager.PaperPluginManagerImpl.callEvent(PaperPluginManagerImpl.java:126) ~[paper-1.20.1.jar:git-Paper-194]
	at org.bukkit.plugin.SimplePluginManager.callEvent(SimplePluginManager.java:615) ~[paper-api-1.20.1-R0.1-SNAPSHOT.jar:?]
	at ia.m.jS.a(SourceFile:401) ~[ItemsAdder_3.5.1.jar:?]
	at ia.m.cm.a(SourceFile:583) ~[ItemsAdder_3.5.1.jar:?]
	at com.destroystokyo.paper.event.executor.MethodHandleEventExecutor.execute(MethodHandleEventExecutor.java:40) ~[paper-api-1.20.1-R0.1-SNAPSHOT.jar:?]
	at co.aikar.timings.TimedEventExecutor.execute(TimedEventExecutor.java:81) ~[paper-api-1.20.1-R0.1-SNAPSHOT.jar:git-Paper-194]
	at org.bukkit.plugin.RegisteredListener.callEvent(RegisteredListener.java:70) ~[paper-api-1.20.1-R0.1-SNAPSHOT.jar:?]
	at io.papermc.paper.plugin.manager.PaperEventManager.callEvent(PaperEventManager.java:54) ~[paper-1.20.1.jar:git-Paper-194]
	at io.papermc.paper.plugin.manager.PaperPluginManagerImpl.callEvent(PaperPluginManagerImpl.java:126) ~[paper-1.20.1.jar:git-Paper-194]
	at org.bukkit.plugin.SimplePluginManager.callEvent(SimplePluginManager.java:615) ~[paper-api-1.20.1-R0.1-SNAPSHOT.jar:?]
	at ia.m.cu.c(SourceFile:217) ~[ItemsAdder_3.5.1.jar:?]
	at org.bukkit.craftbukkit.v1_20_R1.scheduler.CraftTask.run(CraftTask.java:101) ~[paper-1.20.1.jar:git-Paper-194]
	at org.bukkit.craftbukkit.v1_20_R1.scheduler.CraftScheduler.mainThreadHeartbeat(CraftScheduler.java:480) ~[paper-1.20.1.jar:git-Paper-194]
	at net.minecraft.server.MinecraftServer.tickChildren(MinecraftServer.java:1479) ~[paper-1.20.1.jar:git-Paper-194]
	at net.minecraft.server.dedicated.DedicatedServer.tickChildren(DedicatedServer.java:446) ~[paper-1.20.1.jar:git-Paper-194]
	at net.minecraft.server.MinecraftServer.tickServer(MinecraftServer.java:1393) ~[paper-1.20.1.jar:git-Paper-194]
	at net.minecraft.server.MinecraftServer.runServer(MinecraftServer.java:1170) ~[paper-1.20.1.jar:git-Paper-194]
	at net.minecraft.server.MinecraftServer.lambda$spin$0(MinecraftServer.java:317) ~[paper-1.20.1.jar:git-Paper-194]
	at java.lang.Thread.run(Thread.java:833) ~[?:?]
```

### Problematic items yml configuration file (optional)

```shell
info:
  namespace: custom_block
items:
  test:
    display_name: test
    resource:
      generate: true
      material: PAPER
      textures:
        - test.png
    specific_properties:
      block:
        placed_model:
          type: REAL_NOTE
          break_particles: ITEM
        cancel_drop: true
```

### Other files, you can drag and drop them here to upload. (optional)

This all the plugin needed to reproduce the issue. Jar of BetonQuest is special that the dev give me to see error message when breaking the block. If you want the normal one please download the dev build on this [link](https://docs.betonquest.org/2.0-DEV/Downloads/?path=org/betonquest/betonquest/2.0.0-SNAPSHOT/betonquest-2.0.0-20231110.210356-648-shaded.jar&filename=BetonQuest.jar).
[iabug.zip](https://github.com/PluginBugs/Issues-ItemsAdder/files/13326228/iabug.zip)

### Screenshots/Videos (you can drag and drop files or paste links)

https://github.com/PluginBugs/Issues-ItemsAdder/assets/88344458/d19bf31c-cffd-4123-a28b-94796589470a
![Discord_z7DkhWmJIe](https://github.com/PluginBugs/Issues-ItemsAdder/assets/88344458/2b5257dc-7117-4fcc-b21e-41ee6f1cca18)

**Comment by LoneDev6** [comment]
BlockBreakEvent is fired when I want to check if the player can break a block in a particular location or not.

```java
    public static boolean canBreak(Block block, Player player)
    {
        if(block.getType() == Material.AIR)
            return false;
        FakeBlockBreakEvent b = new FakeBlockBreakEvent(block, player);
        b.setDropItems(false);
        Bukkit.getServer().getPluginManager().callEvent(b);
        return !b.isCancelled();
    }
```

Anyway you're not supposed to rely on the `BlockPlaceEvent` for custom blocks.
The `BlockPlaceEvent` is triggered to check for permissions, as you can see if you run `e.getEventName()` it won't output `BlockPlaceEvent` because it's a fake event used for that purpose.
You can also see it's type is set to `STONE`.

You should use my custom event: [CustomBlockPlaceEvent](https://github.com/LoneDev6/API-ItemsAdder/blob/master/src/main/java/dev/lone/itemsadder/api/Events/CustomBlockPlaceEvent.java)


A lot of plugins use this method to check for build permissions, so you should really check if the `BlockPlaceEvent` is real or not, this is not something only ItemsAdder is doing.


```java
    /**
     * Check if an event is a real Bukkit event or not.
     * Useful to skip fake events used by some plugins as permission checking hack (example: ItemsAdder, mcMMO and other).
     *
     * @param e   The event to check.
     * @param <T> The event clazz.
     * @return true if fake, false if not.
     */
    public static <T extends BlockPlaceEvent> boolean isFake(T e)
    {
        Class<? extends BlockPlaceEvent> clazz = e.getClass();
        return !(clazz == BlockPlaceEvent.class || clazz == BlockMultiPlaceEvent.class);
    }
```


example:
```java
  @EventHandler
    void example(BlockPlaceEvent e)
    {
        if (Events.isFake(e))
            return;
```


More info: https://github.com/PluginBugs/Issues-ItemsAdder/issues/2352#issuecomment-1468109696

---

Let me know what you think, you can also send this page to the dev of BetonQuest so that we can discuss about solutions.

**Comment by Wolf2323** [comment]
Hey, i am one of the developers of BetonQuest.
I just read this and i have one question before may continue in any way.
In what priority do you listen to the event? in BetonQuest it looks like this:
```
    @EventHandler(ignoreCancelled = true, priority = EventPriority.HIGHEST)
```

**Comment by LoneDev6** [comment]
In this particular case event priority is not important as the event won't be cancelled in time before reaching your event listener.
So you have to use my little utility to make sure the event is the Bukkit event and not a fake event created by plugins like mine.
Check the code I provided, add it to your plugin, test and you'll see it will catch only Bukkit block break events.

Let me know and thanks for the answer

**Comment by Wolf2323** [comment]
I believe I now understand what your plugin is doing. It appears that when a BlockBreakEvent is triggered, your plugin calls a FakeBlockBreakEvent to check permissions. It's worth noting that a FakeBlockBreakEvent essentially behaves like a regular BlockBreakEvent.

The consequence of this approach is that any plugin not explicitly compatible with yours will receive two BlockBreakEvent events. This can lead to issues in the logic of such plugins, particularly when they perform actions like block counting or block logging, as is common in many logging plugins.

For instance, in the case of BetonQuest, addressing this issue would require adding support for your plugin. However, it's important to mention that there is currently no interest or intention to add explicit support for ItemsAdder.

It's worth pointing out that this usage of the Event API is not the intended or recommended way to handle events. This situation appears to be an issue within your plugin's design rather than something that external developers can easily address or fix.

**Comment by LoneDev6** [comment]
Yes I totally see that point, but there is no clean way to make sure that a player has permission to place a block on a particular location ondemand, this is something the Spigot API lacks and I sincerely don't want to hook every single protection plugin to check if the player can place/break a block.
My solution is to simulate the events and check their cancellation state which was inspired by McMMO method.
https://github.com/mcMMO-Dev/mcMMO/pull/4296

For example, checking if a player can damage an entity:
https://github.com/mcMMO-Dev/mcMMO/blob/89b895af81285465eab256f467809c257eac21c1/src/main/java/com/gmail/nossr50/util/skills/CombatUtils.java#L576

canDamage function:
https://github.com/mcMMO-Dev/mcMMO/blob/89b895af81285465eab256f467809c257eac21c1/src/main/java/com/gmail/nossr50/util/skills/CombatUtils.java#L888

**Comment by LoneDev6** [comment]
I can't do that simply because I need to check if players can break blocks outside of event listeners.
For example my hammer item which breaks blocks 3x3, I have to check if the player can break the surrounding blocks which are not related to the current center block provided by the BlockBreakEvent.
To do so I call fake BlockBreakEvents and check if they got cancelled by a protection plugin.

I hope it's clear, I know it's tricky but I've seen people doing that trick since Bukkit era.

**Comment by Wolf2323** [comment]
I appreciate the clarification of your use case. It's clear that you need to check if players can break blocks outside of event listeners, especially when handling actions like the hammer item breaking multiple blocks in a 3x3 pattern.

In light of your explanation, one potential improvement could be to have the hammer item call the fake event only for the surrounding blocks while excluding the original one. This approach might allow for a more seamless integration with the event system while still achieving your desired functionality. Therefore, my suggestion may still hold value in optimizing the event handling process.

**Comment by LoneDev6** [comment]
There is an underlying issue where your plugin won't be able to know if the block was actually broken or not, because the fake event wouldn't be cancelled.

**Comment by Wolf2323** [comment]
I understand your concern about the underlying issue of not being able to determine whether a block was actually broken or not due to the fake event not being cancelled.

In this context, it raises a broader question: What do you expect other plugins, which may encounter issues with this changed behavior, to do? From my perspective, your example seems to make sense primarily when the intention is to break all blocks. However, if you were to accept that only breakable blocks should be affected, the solution i described should theoretically work. The only change would be in the behavior of your hammer item, where some blocks might not break.

It's worth noting that you've already presented two potential solutions, but both require knowledge about your plugin and security checks for events, which ideally shouldn't be necessary.

**Comment by LoneDev6** [comment]
Can you please check if the issue was fixed? I should have fixed it indirectly by fixing this issue: #3155
